packages/monitor-v2/src/monitor-polymarket/common.ts

```diff
@@ -0,0 +1,422 @@
+  liquidityNum: number;
+  volumeNum: number;
+  clobTokenIds: [string, string];
+}
+
+interface PolymarketMarketWithAncillaryData extends PolymarketMarket {
+  ancillaryData: string;
+}
+
+export interface PolymarketWithEventData extends PolymarketMarketWithAncillaryData {
+  txHash: string;
+  requester: string;
+  proposer: string;
+  timestamp: number;
+  expirationTimestamp: number;
+  eventTimestamp: number;
+  eventBlockNumber: number;
+  eventIndex: number;
+  proposalTimestamp: number;
+  identifier: string;
+  ancillaryData: string;
+  proposedPrice: string;
+}
+
+interface PolymarketOrderBook {
+  market: string;
+  asset_id: string;
+  bids: { price: string; size: string }[];
+  asks: { price: string; size: string }[];
+  hash: string;
+}
+
+export type OrderBookSide = { price: number; size: number }[];
+export interface MarketOrderbooks {
+  orderBooks: [{ bids: OrderBookSide; asks: OrderBookSide }, { bids: OrderBookSide; asks: OrderBookSide }];
+}
+
+export interface StoredNotifiedProposal {
+  txHash: string;
+  question: string;
+  proposedPrice: string;
+  notificationTimestamp: number;
+}
+
+// Helper function to process markets in chunks
+const processMarketsInChunks = async (
+  markets: PolymarketWithEventData[],
+  chunkSize: number,
+  callback: (m: PolymarketWithEventData[]) => Promise<void>
+) => {
+  for (let i = 0; i < markets.length; i += chunkSize) {
+    const chunk = markets.slice(i, i + chunkSize);
+    await callback(chunk);
+    await sleep(500); // Introduce a 0.5 second delay between each chunk
+  }
+};
+
+export const formatPriceEvents = async (
+  events: ProposePriceEvent[]
+): Promise<
+  {
+    txHash: string;
+    requester: string;
+    proposer: string;
+    timestamp: number;
+    eventTimestamp: number;
+    eventBlockNumber: number;
+    expirationTimestamp: number;
+    proposalTimestamp: number;
+    identifier: string;
+    ancillaryData: string;
+    proposedPrice: string;
+    eventIndex: number;
+  }[]
+> => {
+  const ooDefaultLiveness = 7200;
+  return Promise.all(
+    events.map(async (event: ProposePriceEvent) => {
+      const block = await event.getBlock();
+      return {
+        txHash: event.transactionHash,
+        requester: event.args.requester,
+        proposer: event.args.proposer,
+        timestamp: event.args.timestamp.toNumber(),
+        eventTimestamp: block.timestamp,
+        eventBlockNumber: event.blockNumber,
+        expirationTimestamp: event.args.expirationTimestamp.toNumber(),
+        proposalTimestamp: event.args.expirationTimestamp.toNumber() - ooDefaultLiveness,
+        identifier: event.args.identifier,
+        ancillaryData: event.args.ancillaryData,
+        proposedPrice: ethers.utils.formatEther(event.args.proposedPrice),
+        eventIndex: event.logIndex,
+      };
+    })
+  );
+};
+
+export const getPolymarketMarkets = async (params: MonitoringParams): Promise<PolymarketMarket[]> => {
+  const whereClause =
+    " created_at > '2023-01-01'" +
```

Comment:
I don't think a creation time filter is really ever safe, because these markets can last for years (there's already a trump 2024 market).

I think there are ways to make the performance impact almost negligible even if we query every market:
  a. Query them in parallel. I think right now, we're doing multicalls one after the other. We could use a Promise.all to do these much faster. Our node providers can handle it. Note: if you want to limit parallelization due to rate limits or networking limits on the calling machine, we can use bluebird promises to limit the number that are called in parallel.
  b. To avoid making tons of repetitive calls in a serverless env, we could do some redis caching, which shouldn't be too much complexity or code. I think a is probably fine for now, but if we notice this using a lot of quota, caching is an easy option.

It also looks like you have done some work to filter down the number of markets we actually need to check. Will look more closely at that!